modules/core/client/services/menu.client.service.js

```diff
@@ -0,0 +1,196 @@
+
+      // Push new menu item
+      service.menus[menuId].items.push({
+        title: options.title || '',
+        state: options.state || '',
+        type: options.type || 'item',
+        class: options.class,
+        roles: ((options.roles === null || typeof options.roles === 'undefined') ? service.defaultRoles : options.roles),
+        position: options.position || 0,
+        items: [],
+        shouldRender: shouldRender
+      });
+
+      // Add submenu items
+      if (options.items) {
+        for (var i in options.items) {
+          if (options.items.hasOwnProperty(i)) {
+            service.addSubMenuItem(menuId, options.state, options.items[i]);
+          }
+        }
+      }
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    // Add submenu item object
+    function addSubMenuItem(menuId, parentItemState, options) {
+      options = options || {};
+
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Search for menu item
+      for (var itemIndex in service.menus[menuId].items) {
+        if (service.menus[menuId].items[itemIndex].state === parentItemState) {
+          // Push new submenu item
+          service.menus[menuId].items[itemIndex].items.push({
+            title: options.title || '',
+            state: options.state || '',
+            roles: ((options.roles === null || typeof options.roles === 'undefined') ? service.menus[menuId].items[itemIndex].roles : options.roles),
+            position: options.position || 0,
+            shouldRender: shouldRender
+          });
+        }
+      }
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    // Get the menu object by menu id
+    function getMenu(menuId) {
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    function init() {
+      // A private function for rendering decision
+      shouldRender = function (user) {
+        if (!!~this.roles.indexOf('*')) {
+          return true;
+        } else {
+          if (!user) {
+            return false;
+          }
+
+          for (var userRoleIndex in user.roles) {
+            if (user.roles.hasOwnProperty(userRoleIndex)) {
+              for (var roleIndex in this.roles) {
+                if (this.roles.hasOwnProperty(roleIndex) && this.roles[roleIndex] === user.roles[userRoleIndex]) {
+                  return true;
+                }
+              }
+            }
+          }
+        }
+
+        return false;
+      };
+
+      // Adding the topbar menu
+      addMenu('topbar', {
+        roles: ['*']
+      });
+    }
+
+    // Remove existing menu object by menu id
+    function removeMenu(menuId) {
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Return the menu object
+      delete service.menus[menuId];
```

Comment:
Not sure why it matters - the method is referenced no where in code.  Additionally, if it's used, it's not going to be used to the point where performance is an issue.  Also, if someone was to iterate over service.menus, then it would be faster if the key was not there.